presto-geospatial/src/main/java/com/facebook/presto/geospatial/GeoFunctions.java

```diff
@@ -0,0 +1,513 @@
+            Polygon polygon = (Polygon) geometry.getEsriGeometry();
+            return polygon.getPointCount();
+        }
+        else if (geometry.geometryType().equals(MULTI_POINT)) {
+            return ((MultiPoint) geometry.getEsriGeometry()).getPointCount();
+        }
+        return ((MultiPath) geometry.getEsriGeometry()).getPointCount();
+    }
+
+    @SqlNullable
+    @Description("Returns TRUE if and only if the line is closed and simple")
+    @ScalarFunction("ST_IsRing")
+    @SqlType(StandardTypes.BOOLEAN)
+    public static Boolean stIsRing(@SqlType(NAME) Slice input)
+    {
+        OGCGeometry geometry = getGeometry(deserialize(input), "ST_IsRing", LINE_STRING);
+        OGCLineString line = (OGCLineString) geometry;
+        return line.isClosed() && line.isSimple();
+    }
+
+    @Description("Returns the first point of a LINESTRING geometry as a Point")
+    @ScalarFunction("ST_StartPoint")
+    @SqlType(NAME)
+    public static Slice stStartPoint(@SqlType(NAME) Slice input)
+    {
+        OGCGeometry geometry = getGeometry(deserialize(input), "ST_StartPoint", LINE_STRING);
+        MultiPath lines = (MultiPath) geometry.getEsriGeometry();
+        SpatialReference reference = geometry.getEsriSpatialReference();
+        return serialize(OGCGeometry.createFromEsriGeometry(lines.getPoint(0), reference));
+    }
+
+    @Description("Returns the last point of a LINESTRING geometry as a Point")
+    @ScalarFunction("ST_EndPoint")
+    @SqlType(NAME)
+    public static Slice stEndPoint(@SqlType(NAME) Slice input)
+    {
+        OGCGeometry geometry = getGeometry(deserialize(input), "ST_EndPoint", LINE_STRING);
+        MultiPath lines = (MultiPath) geometry.getEsriGeometry();
+        SpatialReference reference = geometry.getEsriSpatialReference();
+        return serialize(OGCGeometry.createFromEsriGeometry(lines.getPoint(lines.getPointCount() - 1), reference));
+    }
+
+    @Description("Return the X coordinate of the point")
+    @ScalarFunction("ST_X")
+    @SqlType(StandardTypes.DOUBLE)
+    public static double stX(@SqlType(NAME) Slice input)
+    {
+        OGCGeometry geometry = getGeometry(deserialize(input), "ST_X", POINT);
+        return ((OGCPoint) geometry).X();
+    }
+
+    @Description("Return the Y coordinate of the point")
+    @ScalarFunction("ST_Y")
+    @SqlType(StandardTypes.DOUBLE)
+    public static double stY(@SqlType(NAME) Slice input)
+    {
+        OGCGeometry geometry = getGeometry(deserialize(input), "ST_Y", POINT);
+        return ((OGCPoint) geometry).Y();
+    }
+
+    @Description("Returns the closure of the combinatorial boundary of this Geometry")
+    @ScalarFunction("ST_Boundary")
+    @SqlType(NAME)
+    public static Slice stBoundary(@SqlType(NAME) Slice input)
+    {
+        OGCGeometry geometry = deserialize(input);
+        OGCGeometry bound = geometry.boundary();
+        if (bound.geometryType().equals(MULTI_LINE_STRING) && ((OGCMultiLineString) bound).numGeometries() == 1) {
```

Comment:
This logic seems unnecessary. I noticed that serialize function outputs 3 pieces of data: (1) SRID, (2) geometry type, (3) ESRI Shape File. However, deserialize function uses only (1) and (3) and implicitly derives (2) from (3). serialize-deserialize cycle applied to a MultiLineString of one LineString produces the same result as converting MultiLineString to a LineString first. Hence, two suggestions:

(1) remove this logic
(2) update serialize to not output shape type